### PR TITLE
Move xds API code into a class, so that common state can be cached.

### PR DESCRIPTION
**Pull request by markdroth**
There are no functional changes here; this is just a small refactoring to prepare for a subsequent PR that will add some additional logging.

### CODE REVIEW

File: src/core/ext/filters/client_channel/xds/xds_api.cc [hunk 1]
```diff
@@ -279,183 +252,106 @@ grpc_slice XdsLdsRequestCreateAndEncode(const std::string& server_name,
...
-  // Set type_url.
-  envoy_api_v2_DiscoveryRequest_set_type_url(request,
-                                             upb_strview_makez(kRdsTypeUrl));
-  // Set nonce.
-  if (!nonce.empty()) {
-    envoy_api_v2_DiscoveryRequest_set_response_nonce(
-        request, upb_strview_makez(nonce.c_str()));
-  }
-  // Set error_detail if it's a NACK.
-  if (error != GRPC_ERROR_NONE) {
```
Comment (donnadionne) [inline src/core/ext/filters/client_channel/xds/xds_api.cc]: Maybe i am missing something, but i feel like from 329-341 this error handling is not included in the refactored code?

Comment (donnadionne) [reply src/core/ext/filters/client_channel/xds/xds_api.cc]: should it be added to CreateDiscoveryRequest ?

Comment (markdroth) [reply src/core/ext/filters/client_channel/xds/xds_api.cc]: As per our discussion, it is actually present in `CreateDiscoveryRequest()`.  It's just that the way github displays diffs makes it a little non-obvious.

**Comment by markdroth** [comment]
Known issues: #15027 #20468 #20609

**Pull request merged**